### PR TITLE
Allow custom probability distributions

### LINKED ISSUE #19057: Allowing users to make custom probability distributions
As of now, there are plenty of probability distributions defined already in the package, for example [here](https://github.com/sympy/sympy/blob/master/sympy/stats/crv_types.py). We can add more such distributions to it but that will just lead to a forever increasing code size. To avoid this probably we can allow users to inherit some classes from `sympy.stats` to make their own custom distributions. I am not sure if this is already allowed but if not then following can be done to expose the framework of distribution types,

1. Bring some of the classes like, `SingleContinuousDistribution`, etc to `sympy.stats` namespace so that they can be imported and inherited in the custom distribution.
2. Add some examples in the documentation of the process of doing this.

Comments from community are needed to think and decide on this.

### PR DESCRIPTION
**Pull request by Smit-create**
Allow user to define their probability distributions
#### References to other Issues or PRs

Fixes #19057

#### Other comments
- [x] Define all HandmadeDistributions classes and move to `sympy.stats` namespace
- [x] Add missing tests of Handmade Distributions and documentation

#### Release Notes

* stats
   * Allowing the user to define custom distributions

### CODE REVIEW

File: sympy/stats/drv_types.py [hunk 1]
```diff
@@ -49,6 +50,64 @@ def rv(symbol, cls, *args):
...
+
+    Examples
+    ========
+
+    >>> from sympy.stats import DiscreteRV
+    >>> from sympy import Rational, Symbol
+    >>> x = Symbol('x')
+    >>> n = 10
+    >>> density = Rational(1, 10)
+    >>> X = DiscreteRV(x, density, set=set(range(n)))
```
Comment (czgdp1807) [inline sympy/stats/drv_types.py:92]: A better way can be to pass `pmf` as a `dict`(or whatever the type used to store `pmf` inside `sympy.stats`).

Comment (Smit-create) [reply sympy/stats/drv_types.py:92]: Actually, we use `pdf` for discrete distributions, with `pdf` as an expression. And `dict` is passed as `pmf` in Finite distributions.

Comment (czgdp1807) [reply sympy/stats/drv_types.py:91]: > And `dict` is passed as `pmf` in Finite distributions.

I see, can we pass `dict` as `pmf` to Finite distributions?

Comment (Smit-create) [reply sympy/stats/drv_types.py:91]: Yes
https://github.com/sympy/sympy/blob/6daf556517f9fa61a7805fe93d8a366688781617/sympy/stats/frv_types.py#L74-L96

**Comment by czgdp1807** [comment]
LGTM. We can go for this if no one objects for some time.

**Comment by oscarbenjamin** [comment]
I just tried creating a symbolically-sized discrete random variable and got this:
```julia
In [1]: from sympy.stats import E, P, DiscreteRV                                                                                  

In [2]: x = Symbol('x')                                                                                                           

In [3]: n = Symbol('n')                                                                                                           

In [4]: X = DiscreteRV(x, 1/n, Range(n))                                                                                          
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-6b398ebcef34> in <module>
----> 1 X = DiscreteRV(x, 1/n, Range(n))

~/current/sympy/sympy/sympy/stats/drv_types.py in DiscreteRV(symbol, density, set)
    102     """
    103     set = sympify(set)
--> 104     pdf = Piecewise((density, set.as_relational(symbol)), (0, True))
    105     pdf = Lambda(symbol, pdf)
    106     return rv(symbol.name, DiscreteDistributionHandmade, pdf, set)

~/current/sympy/sympy/sympy/sets/fancysets.py in as_relational(self, x)
    918         """Rewrite a Range in terms of equalities and logic operators. """
    919         from sympy.functions.elementary.integers import floor
--> 920         if self.size == 1:
    921             return Eq(x, self[0])
    922         else:

~/current/sympy/sympy/sympy/sets/fancysets.py in size(self)
    725             if dif.has(Symbol) or self.step.has(Symbol) or (
    726                     not self.start.is_integer and not self.stop.is_integer):
--> 727                 raise ValueError('invalid method for symbolic range')
    728         if dif.is_infinite:
    729             return S.Infinity

ValueError: invalid method for symbolic range
```
Is that something that can be handled as well?

**Comment by Smit-create** [comment]
> Is that something that can be handled as well?

That seems to be an issue with `sets`.

**Comment by oscarbenjamin** [comment]
It looks like `Range` needs a bit of work but are there any examples of finite but symbolically sized distributions that this PR can work with?

**Comment by Smit-create** [comment]
That is because we return `Density` object for symbolic parameters.
https://github.com/sympy/sympy/blob/22c57c73758503b2ae857258549232351cca3861/sympy/stats/frv.py#L197-L200

I think these all will be resolved once some issues with `Range` are fixed.

**Comment by czgdp1807** [comment]
> once some issues

Can you please investigate what are the specific issues in `Range` due to which custom RVs aren't being created? Make a comment here and reference the same in a new issue so that we don't loose track of the problems in `Range`.

**Comment by Smit-create** [comment]
```python
>>> Range(n).size
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/smit/Smit Lunagariya/sympy/sympy/sets/fancysets.py", line 727, in size
    raise ValueError('invalid method for symbolic range')
ValueError: invalid method for symbolic range
```
I see, that the the size of symbolic range is not defined and hence while defining RV on the given set, it is unable to convert it into relational set.
This can also be noticed with,
```python
>>> Sum(1/n, Range(n)).doit()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/smit/Smit Lunagariya/sympy/sympy/concrete/summations.py", line 158, in __new__
    obj = AddWithLimits.__new__(cls, function, *symbols, **assumptions)
  File "/home/smit/Smit Lunagariya/sympy/sympy/concrete/expr_with_limits.py", line 494, in __new__
    pre = _common_new(cls, function, *symbols, **assumptions)

  File "/home/smit/Smit Lunagariya/sympy/sympy/concrete/expr_with_limits.py", line 48, in _common_new
    limits, orientation = _process_limits(*symbols)
  File "/home/smit/Smit Lunagariya/sympy/sympy/concrete/expr_with_limits.py", line 107, in _process_limits
    if len(V) == 2 and isinstance(V[1], Range):
  File "/home/smit/Smit Lunagariya/sympy/sympy/sets/fancysets.py", line 714, in __len__
    rv = self.size
  File "/home/smit/Smit Lunagariya/sympy/sympy/sets/fancysets.py", line 727, in size
    raise ValueError('invalid method for symbolic range')
ValueError: invalid method for symbolic range
>>> Sum(1/n, (x,Range(n))).doit() # This works
1
```

**Comment by czgdp1807** [comment]
It appears that `FiniteDistributionHandmade` needs some changes for allowing symbolically sized distributions. This needs to be handled in a different PR and is out of scope of this PR.

**Comment by czgdp1807** [comment]
If anyone is having any suggestions other than symbolically sized distributions then please comment.

**Comment by czgdp1807** [comment]
Merging. Thanks.

**Pull request merged**